Avoid breaking current directory on neovim

hkupty opened the pull request:
Neovim allows a `tab-local` directory. The previous directory change function was breaking tab local directory configuration, changing it with `cd`.

This PR fixes the problem.

jreybert commented:
Your patch is not compatible with vim, because of haslocaldir(-1, 0) (Travis arguing)

hkupty commented:
Can we merge this?

jreybert commented on autoload/magit/utils.vim:
```diff
@@ -20,12 +20,43 @@ function! magit#utils#is_submodule(dirname)
+
+" magit#utils#pushd_root goes to git root directory, saving current directory
+" prior to changing the cwd.
+function! magit#utils#pushd_root()
```
`pushd` and and `popd` are not reentrant. Please add a counter in `pushd`, such as:

```
function pushd()
  b:pushd_cnt++
  if ( b:pushd_cnt <= 0 )
    raise error
  else if ( b:pushd_cnt == 1)
    b:_vimagit_cwd = cwd
    chdir root

function popd()
  b:pushd_cnt--
  if ( b:pushd_cnt < 0 )
    raise error
  else if ( b:pushd_cnt == 0 )
    chdir b:_vimagit_cwd
```

jreybert commented on autoload/magit/utils.vim:
```diff
@@ -20,12 +20,43 @@ function! magit#utils#is_submodule(dirname)
+  if has('nvim')
+    let chdir = haslocaldir() ? 'lcd' : haslocaldir(-1, 0) ? 'tcd' : 'cd'
+  else
+    let chdir = exists('*haslocaldir') && haslocaldir() ? 'lcd' : 'cd'
```
`has(nvim)` and `exists('*haslocaldir')` should be evaluated at plugin initialization (outside of a function).

hkupty replied:
We can externalize both checks, but that would only make a difference if we define a function for each case. Otherwise we're swapping a strong check (i.e. `has('nvim')`) that won't change during runtime to something that can be overriden (i.e. `g:magit_running_on_neovim=1`).

By conditioning the implementation like that, we end up with an unmaintainable version of it:

``` vim
if has('nvim')
  function! magit#utils#chdir(dir)
    let chdir = haslocaldir() ? 'lcd' : haslocaldir(-1, 0) ? 'tcd' : 'cd'
    exec chdir a:dir
  endfunction
elseif exists('*haslocaldir')
function! magit#utils#chdir(dir)
    let chdir = haslocaldir() ? 'lcd' : 'cd'
    exec chdir a:dir
  endfunction
else
function! magit#utils#chdir(dir)
    cd a:dir
  endfunction
endif
```

jreybert replied:
Why do you think it is unmaintainable? I think this is a good solution, even better than my previous solution of storing the chdir command in a variable.

I do prefer this solution. Maybe this is a micro optimization, but if vimagit wins two if statements during runtime, I prefer that.

Several points here:
- `has('nvim')` is not enough check, you should also check nvim version, because not all nvim version has tcd.
- your integration has failed because of #89 : now, `a:dir` must be escaped with `fnameescape(a:dir)`

hkupty replied:
With this solution, we end up with 3 functions to maintain, so we have three places to add `fnameescape` to.

Also,this is a premature optimization; we're trading readability and maintainability for _theoretical_ performance gains, which never proved to be actually necessary. 

Plus, if we scale the amount of time spent on operations, changing the directory setting of nvim/vim is way faster than making a system call to interact with git, for example, so even if vimagit was extremely slow, this optimization would not perceivable results.

jreybert replied:
Well, I won't fight against this, in the end this is only a micro optim.

One day, I hope changing directory won't be necessary... I already tried to play around with GIT_WORK_TREE and GIT_DIR environment variables, but it was a dead end, I can't remember why.
I should give another try: in case it works, vimagit won't need to change dir anymore.

jreybert replied:
I would prefer to not mix two solutions based on third party presence.
At least, a mix of 2 solutions based on git version would be OK.

hkupty replied:
I realised my comment was stupid. Still, you can use `--git-dir` and `--work-dir`, so you don't need to mess with env vars.

jreybert replied:
I already tried these. I try to remember why I did not went to the end... Maybe because these options appeared lately in git. I'll check that, but it could worth a try to support both:
- Old git version with magit#chdir
- Not so old git version with --git-dir

jreybert commented on autoload/magit/utils.vim:
```diff
@@ -20,12 +20,43 @@ function! magit#utils#is_submodule(dirname)
+" magit#utils#pushd_root goes to git root directory, saving current directory
+" prior to changing the cwd.
+function! magit#utils#pushd_root()
+  if has('nvim')
```
Why don't you use `magit#utils#chdir` here?

jreybert commented on autoload/magit/utils.vim:
```diff
@@ -20,12 +20,43 @@ function! magit#utils#is_submodule(dirname)
+
+  if root !=? cwd
+    call magit#utils#chdir(root)
+    let g:_vimagit_cwd = cwd
```
`_vimait_cwd` variable must be local to the current buffer (user may have several magit session opened in one vim session).

jreybert reviewed (changes requested):
I have multiple comments in this PR:
- please split the tcd fix and the pushd feature in two different PRs
- make the PR for branch next
- for the tcd fix, could you provide me a way to reproduce your error?
- for the pushd feature, please see the comments, and:
  - use it everywhere
  - don't bother of tcd fix in the pushd PR, it will be merged when the two PR will be merged

hkupty committed:
Avoid breaking tab-local directory settings

Neovim has support for tab-local settings, additionally to global and
window-local.

When a tab is set with `tcd`, if the directory is changed with `cd`, the
tab-local setting is discarded.

This patch fixes the current bug by setting checking the granularity of current
setting prior to changing directory.

jreybert commented on autoload/magit/utils.vim:
```diff
@@ -55,7 +60,7 @@ endfunction
 	let dir = getcwd()
 	try
-		call magit#utils#lcd(magit#git#top_dir())
+		execute magit#utils#chdir(magit#git#top_dir())
```
Can you explain why you changed `call` to `execute`? And why don't you use `execute` everywhere in that case?

hkupty replied:
This was not supposed to happen. Will revert to `call`.

jreybert reviewed (commented):
This is close to be merged, thanks for your hard work and your patience!

hkupty commented on autoload/magit/utils.vim:
```diff
@@ -20,12 +20,43 @@ function! magit#utils#is_submodule(dirname)
+  if has('nvim')
+    let chdir = haslocaldir() ? 'lcd' : haslocaldir(-1, 0) ? 'tcd' : 'cd'
+  else
+    let chdir = exists('*haslocaldir') && haslocaldir() ? 'lcd' : 'cd'
```
You could try to find `rgit` on the local path. If it's there, then just point to the directory root and it simply works.

hkupty commented:
I'll probably need your help to write a test here.
Here is how to reproduce the bug:

``` vim
" Open neovim
tcd /tmp/git1
tabnew
tcd /tmp/git2
tabprevious
echo getcwd() "== /tmp/git1
Magit
tabnext
echo getcwd() "== /tmp/git2
Magit
tabprevious
echo getcwd() "== /tmp/git2
```

jreybert commented on autoload/magit/utils.vim:
```diff
@@ -20,12 +20,17 @@ function! magit#utils#is_submodule(dirname)
+  " This is a dirty hack to fix tcd breakages on neovim.
+  " Future work should be based on nvim API and has a `if has('nvim')` check.
+  if has('nvim')
+    let chdir = haslocaldir() ? 'lcd' : haslocaldir(-1, 0) ? 'tcd' : 'cd'
```
`:tcd` has been introduced right after nvim 0.1.3 (I did not check, it should be officially present since nvim 0.1.4).

Before `:tcd` introduction, `haslocaldir(-1, 0)` will return an error:

```
E118: Too many arguments for function: haslocaldir                                                                                                                                                                                                                                
E15: Invalid expression: haslocaldir(-1, 0)                                                                                                                                                                                                                                       
```

hkupty replied:
Will do. I think nvim allows checking for version with `has('nvim-0.1.4')`. If not I'll probably be able to `exists('*tcd')`.

jreybert replied:
I prefer exists() solution .

hkupty replied:
Yeah, I already implemented using `exists`. Github doesn't update diffs properly now... :/

jreybert commented:
Good job 👍
